### PR TITLE
Add UpdateProductSuppliers command

### LINKED ISSUE #19683: [Options tab] - Suppliers 5/5
As a merchant, I want to be able to manage the options of my products.

**Suppliers**: This part is displayed only if you have created and enabled at least one supplier. Indicating the product's supplier is not really important for customers, but it may turn out to be an essential part for the merchant’s internal management, not least when managing stock: the merchant simply needs to know who he bought the product from.
  * **Information message about suppliers**: “This interface allows you to specify the suppliers of the current product and its combinations if any. You can specify supplier references according to previously associated suppliers.” Click on read more link to display the following information: “When using the advanced stock management tool \(see Shop Parameters &gt; Products settings\), the values you define \(price, references\) will be used in supply orders.”
  * **Choose the suppliers associated with this product**: All enabled suppliers are displayed with a checkbox before their names. If it’s checked, the product is associated to the supplier and displayed on the supplier page in front-office.
  * **Default supplier**: When you check a supplier, a radio button allowing to choose the default supplier appears. When there is only one supplier checked, the default supplier radio button is automatically checked. The first supplier checked is the default supplier. When there are many suppliers checked, you can change the default supplier.
  * **Supplier reference\(s\)**: This part is displayed only if you have associated at least one supplier to the product. The supplier's section also features a table that enables you to set the precise reference and unit price/currency for each product combination, per supplier. If the product has more than one supplier, the table will display each supplier one after the other.
  * **Information message about supplier reference\(s\)**: “You can specify product reference\(s\) for each associated supplier. Click "Save" after changing selected suppliers to display the associated product references.”
  * **Table of suppliers**:
    * **The product name**: Display the product name
    * **Supplier reference**: Field to specify the product reference for each supplier. 
    * **Price \(tax excl.\)**: Field to specify the product price for each supplier. (check wording https://github.com/PrestaShop/PrestaShop/issues/18701)
    * **Currency**: Dropdown with all installed and enabled currencies. If there are many currencies, the default one is selected by default. Allow to specify the product currency for each supplier.

---

Edited by Louise B. on September 21st:

'Disassociate one supplier should delete only this supplier's data and not the other suppliers' data. Same for the deletion of a supplier in Catalog > Brands & Suppliers > Suppliers.' (https://github.com/PrestaShop/prestashop-specs/pull/132)

### PR DESCRIPTION
**Pull request by zuk3975**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop 
| Description?  | `UpdateProductSuppliersHandler` contains three handlers inside: `AddProductSupplierHandler`, `UpdateProductSupplierHandler` and `DeleteProductSupplierHandler`. It decides what to add/update/delete depending on provided suppliers list. Also added `ProductSupplierOptions` in ProductForEditing query result and covered basic usecases with behats (including some combination asserts - had to move combination creation(in tests) to `CombinationFactory` for reusability)
| Type?         | refacto
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Partially fixes #19683
| How to test?  | none

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/20299)

### CODE REVIEW

File: src/Adapter/Product/CommandHandler/AddProductSupplierHandler.php [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+    /**
+     * @param ProductSupplier $productSupplier
+     * @param AddProductSupplierCommand $command
+     *
+     * @throws CurrencyNotFoundException
+     * @throws ProductNotFoundException
+     * @throws SupplierNotFoundException
+     */
+    private function fillEntityWithCommandData(ProductSupplier $productSupplier, AddProductSupplierCommand $command)
```
Comment (aleksiuno) [inline src/Adapter/Product/CommandHandler/AddProductSupplierHandler.php]: I suppose you could use `void` return type here if possible.

File: src/Adapter/Product/AbstractProductSupplierHandler.php [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+                0,
+                $e
+            );
+        }
+
+        return $productSupplier;
+    }
+
+    /**
+     * @param ProductSupplier $productSupplier
```
Comment (matks) [inline src/Adapter/Product/AbstractProductSupplierHandler.php:73]: phpDoc is outdated

File: src/Adapter/Product/CommandHandler/AddProductSupplierHandler.php [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+use PrestaShop\PrestaShop\Core\Domain\Product\Supplier\Exception\ProductSupplierException;
+use PrestaShop\PrestaShop\Core\Domain\Product\Supplier\ValueObject\ProductSupplierId;
+use PrestaShop\PrestaShop\Core\Domain\Supplier\Exception\SupplierNotFoundException;
+use PrestaShopException;
+use Product;
+use ProductSupplier;
+use Supplier;
+
+/**
+ * Handles @var AddProductSupplierCommand using legacy object model
```
Comment (matks) [inline src/Adapter/Product/CommandHandler/AddProductSupplierHandler.php:47]:
```suggestion
 * Handles AddProductSupplierCommand using legacy object model
```

File: src/Adapter/Product/CommandHandler/AddProductSupplierHandler.php [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    /**
+     * @param int $productId
+     * @param int $supplierId
+     * @param int $currencyId
+     *
+     * @throws CurrencyNotFoundException
+     * @throws ProductNotFoundException
+     * @throws SupplierNotFoundException
+     */
+    private function assertRelatedEntitiesExist(int $productId, int $supplierId, int $currencyId): void
```
Comment (matks) [inline src/Adapter/Product/CommandHandler/AddProductSupplierHandler.php:120]: 👍

File: src/Adapter/Product/CommandHandler/DeleteProductSupplierHandler.php [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+use PrestaShop\PrestaShop\Core\Domain\Product\Supplier\Command\DeleteProductSupplierCommand;
+use PrestaShop\PrestaShop\Core\Domain\Product\Supplier\CommandHandler\DeleteProductSupplierHandlerInterface;
+use PrestaShop\PrestaShop\Core\Domain\Product\Supplier\Exception\CannotDeleteProductSupplierException;
+use PrestaShop\PrestaShop\Core\Domain\Product\Supplier\Exception\ProductSupplierException;
+use PrestaShop\PrestaShop\Core\Domain\Product\ValueObject\ProductId;
+use PrestaShopException;
+use ProductSupplier;
+
+/**
+ * Handles @var DeleteProductSupplierCommand using legacy object model
```
Comment (matks) [inline src/Adapter/Product/CommandHandler/DeleteProductSupplierHandler.php:44]:
```suggestion
 * Handles DeleteProductSupplierCommand using legacy object model
```

File: src/Adapter/Product/CommandHandler/DeleteProductSupplierHandler.php [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    {
+        $productId = (int) $productSupplier->id_product;
+        $product = $this->getProduct(new ProductId($productId));
+
+        // check if default supplier was deleted with this command
+        if ((int) $product->id_supplier !== (int) $productSupplier->id_supplier) {
+            return;
+        }
+
+        $product->id_supplier = 0;
```
Comment (matks) [inline src/Adapter/Product/CommandHandler/DeleteProductSupplierHandler.php:89]: Nice ! @PierreRambaud was worried about that (see https://github.com/PrestaShop/PrestaShop/pull/20198/) but you handled it

File: src/Adapter/Product/CommandHandler/UpdateProductSuppliersHandler.php [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+            $command->setCombinationId($productSupplier->getCombinationId());
+        }
+
+        $this->updateProductSupplierHandler->handle($command);
+    }
+
+    /**
+     * @param ProductId $productId
+     * @param int $defaultSupplierId
+     *
```
Comment (matks) [inline src/Adapter/Product/CommandHandler/UpdateProductSuppliersHandler.php:229]: Maybe a little comment in this phpDoc could explain the complex data structure for default supplier (with combinations, without ...) ? to help clear the confusion on this topic :)

File: src/Core/Domain/Product/Customization/CustomizationField.php [hunk 8]
```diff
@@ -64,6 +64,7 @@ class CustomizationField
      * @param bool $required
      * @param int|null $customizationFieldId
      * @param bool $addedByModule
+     * @todo: require all fields except the id.
```
Comment (matks) [inline src/Core/Domain/Product/Customization/CustomizationField.php:67]: TODO
- outdated ?
- still valid but need to do it in another PR ?

File: tests/Integration/Behaviour/Features/Scenario/Product/update_suppliers.feature [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    And single shop shop1 context is loaded
+    And language "language1" with locale "en-US" exists
+    And there is a currency named "currency1" with iso code "USD" and exchange rate of 0.92
+    And there is a currency named "currency3" with iso code "EUR" and exchange rate of 0.63
+
+  Scenario: Update standard product suppliers
+    And I add new supplier supplier1 with following properties:
+      | name                 | my supplier 1                   |
+      | address              | Donelaicio st. 1                |
+      | city                 | Kaunas                          |
```
Comment (matks) [inline tests/Integration/Behaviour/Features/Scenario/Product/update_suppliers.feature:19]: Cameo 😄  ?

File: tests/Integration/Behaviour/Features/Scenario/Product/update_suppliers.feature [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+      | country              | Lithuania                       |
+      | enabled              | true                            |
+      | description          | en-US:just a supplier           |
+      | meta title           | en-US:my supplier nr two        |
+      | meta description     | en-US:                          |
+      | meta keywords        | en-US:sup,2                     |
+      | shops                | [shop1]                         |
+    Given I add product "product1" with following information:
+      | name       | en-US:magic staff                         |
```
Comment (matks) [inline tests/Integration/Behaviour/Features/Scenario/Product/update_suppliers.feature:39]: 🧙

File: tests/Integration/Behaviour/Features/Scenario/Product/update_suppliers.feature [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+    And product product1 should not have a default supplier
+    And product product1 default supplier reference should be empty
+
+  Scenario: Update combination product suppliers
+    Given I add product "product2" with following information:
+      | name       | en-US:regular T-shirt        |
+      | is_virtual | false                        |
+    And product "product2" has following combinations:
+      | reference | quantity | attributes         |
+      | whiteM    | 15       | Size:M;Color:White |
```
Comment (matks) [inline tests/Integration/Behaviour/Features/Scenario/Product/update_suppliers.feature:96]: Looks good like this 👍

**Review by matks — APPROVED** [review]
I provided some suggestion however the PR is valid and very good, I approve it 👍

File: src/Adapter/Product/CommandHandler/AddProductSupplierHandler.php (hunk 3 above)
Comment (matks) [reply src/Adapter/Product/CommandHandler/AddProductSupplierHandler.php:47]: I see you did it everywhere. I thought initially it was a bad copy-paste. But maybe you did it on purpose ?

Comment (zuk3975) [reply src/Adapter/Product/CommandHandler/AddProductSupplierHandler.php:47]: I did it on purpose, this way if you rename the command (refactor through your editor - at least phpstorm), it will be updated in docblock aswell, Although it looks abit strange,I thought its a good idea, tell me if Im wrong :sweat_smile:

File: src/Core/Domain/Product/Customization/CustomizationField.php (hunk 8 above)
Comment (zuk3975) [reply src/Core/Domain/Product/Customization/CustomizationField.php:67]: still valid but need to do it in another PR

**Review by matthieu-rolland — APPROVED** [review]

File: src/Adapter/Product/CommandHandler/AddProductSupplierHandler.php (hunk 3 above)
Comment (zuk3975) [reply src/Adapter/Product/CommandHandler/AddProductSupplierHandler.php:47]: maybe its better to use `@see` instead?

**Comment by matthieu-rolland** [comment]
thanks @zuk3975 !

File: src/Adapter/Product/CommandHandler/AddProductSupplierHandler.php (hunk 3 above)
Comment (matks) [reply src/Adapter/Product/CommandHandler/AddProductSupplierHandler.php:47]: > refactor through your editor - at least phpstorm), it will be updated in docblock aswellrefactor through your editor - at least phpstorm), it will be updated in docblock aswell

I did not know this nice trick. Let's keep it ! thanks !
